pkg/portrait/algorithm/externaljob/jobcontroller/cron_job.go

```diff
@@ -0,0 +1,123 @@
+	"fmt"
+
+	batchv1 "k8s.io/api/batch/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	autoscalingv1alpha1 "github.com/traas-stack/kapacity/apis/autoscaling/v1alpha1"
+	"github.com/traas-stack/kapacity/pkg/util"
+)
+
+type CronJobHorizontal struct {
+	client client.Client
+}
+
+func NewCronJobHorizontal(client client.Client) Horizontal {
+	return &CronJobHorizontal{
+		client: client,
+	}
+}
+
+func (c *CronJobHorizontal) UpdateJob(ctx context.Context, hp *autoscalingv1alpha1.HorizontalPortrait, cfg *autoscalingv1alpha1.PortraitAlgorithmJob) error {
+	cronJobNamespacedName := types.NamespacedName{
+		Namespace: hp.Namespace,
+		Name:      hp.Name,
+	}
+	cronJob := &batchv1.CronJob{}
+	if err := c.client.Get(ctx, cronJobNamespacedName, cronJob); err != nil {
+		if apierrors.IsNotFound(err) {
+			cronJob = &batchv1.CronJob{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace:   cronJobNamespacedName.Namespace,
+					Name:        cronJobNamespacedName.Name,
+					Annotations: cfg.CronJob.Template.Annotations,
+					Labels:      cfg.CronJob.Template.Labels,
+					OwnerReferences: []metav1.OwnerReference{
+						*util.NewControllerRef(hp),
+					},
+				},
+				Spec: cfg.CronJob.Template.Spec,
+			}
+
+			if err := c.client.Create(ctx, cronJob); err != nil {
+				return fmt.Errorf("failed to create CronJob %q: %v", cronJobNamespacedName, err)
+			}
+
+			return nil
+		}
+
+		return fmt.Errorf("failed to get CronJob %q: %v", cronJobNamespacedName, err)
+	}
+
+	if apiequality.Semantic.DeepEqual(cronJob.Spec, cfg.CronJob.Template.Spec) &&
+		ifUnchanged(cronJob.Labels, cfg.CronJob.Template.Labels) &&
+		ifUnchanged(cronJob.Annotations, cfg.CronJob.Template.Annotations) {
+		return nil
+	}
+
+	patch := client.MergeFrom(cronJob.DeepCopy())
+	cronJob.Spec = cfg.CronJob.Template.Spec
+	if err := c.client.Patch(ctx, cronJob, patch); err != nil {
+		return fmt.Errorf("failed to patch CronJob %q: %v", cronJobNamespacedName, err)
+	}
+
+	return nil
+}
+
+func (c *CronJobHorizontal) CleanupJob(ctx context.Context, hp *autoscalingv1alpha1.HorizontalPortrait) error {
+	cronJobNamespacedName := types.NamespacedName{
+		Namespace: hp.Namespace,
+		Name:      hp.Name,
+	}
+
+	cronJob := &batchv1.CronJob{}
+	if err := c.client.Get(ctx, cronJobNamespacedName, cronJob); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to get CronJob %q: %v", cronJobNamespacedName, err)
+	}
+
+	if err := client.IgnoreNotFound(c.client.Delete(ctx, cronJob)); err != nil {
+		return fmt.Errorf("failed to delete CronJob %q: %v", cronJobNamespacedName, err)
+	}
+	return nil
+}
+
+// ifUnchanged Compare newMap and oldMap, if the key or value in newMap changes, return true, otherwise return false
+func ifUnchanged(oldMap, newMap map[string]string) bool {
```

Comment:
This func can be renamed to `func IsMapValueChanged(oldValues, newValues map[string]string) bool` (reverse the logic) and moved to `util/common.go` so that it can be reused.